server/functions/src/index.ts

```diff
@@ -0,0 +1,126 @@
+  subscribedTopics: string[];
+}
+
+// Implementation of the v1 API"s `getCaseStats` method.
+// TODO: replace with direct Firestore access from the client.
+export const getCaseStats = functions
+  .region("europe-west6")
+  .https.onRequest((request, response) => {
+    // A "fake" of the v1 API: we're just going to return a static file that we
+    // happen to have on disk.
+    // TODO: replace this with actual statistics in Firestore.
+    const data = JSON.parse(fs.readFileSync("casestats.json", "utf8"));
+    response.status(200).json(data);
+  });
+
+// Implementation of the v1 API"s `putClientSettings` method.
+// TODO: replace with direct Firestore acccess from the client.
+export const putClientSettings = functions
+  .region("europe-west6")
+  .https.onRequest((request, response) => {
+    const whoClientId = request.header("Who-Client-ID");
+    if (whoClientId === undefined || whoClientId == null) {
+      response.status(400).send("Missing Who-Client-ID header");
+      return;
+    }
+    const whoPlatform = request.header("Who-Platform");
+    if (whoPlatform === undefined || whoPlatform == null) {
+      response.status(400).send("Missing Who-Platform header");
+      return;
+    }
+    let platform = Platform.WEB;
+    if (whoPlatform == Platform[Platform.ANDROID]) {
+      platform = Platform.ANDROID;
+    } else if (whoPlatform == Platform[Platform.IOS]) {
+      platform = Platform.IOS;
+    }
+
+    if (request.method != "POST") {
+      response.status(400).send("Call must be POST request");
+      return;
+    }
+    let isoCountryCode = request.body["isoCountryCode"];
+    if (isoCountryCode === undefined || isoCountryCode == null) {
+      isoCountryCode = "";
+    } else if (
+      // Don"t even run a regex on a very long string.
+      isoCountryCode.length != 2 ||
+      !isoCountryCode.match(COUNTRY_CODE)
+    ) {
+      response.status(400).send("Invalid isoCountryCode");
+      return;
+    }
+    let fcmToken = request.body["token"];
+    if (fcmToken === undefined || fcmToken == "null") {
+      fcmToken = "";
+    }
+    if (fcmToken.length > FCM_TOKEN_MAX_LENGTH) {
+      response.status(400).send("Invalid FCM Token");
+      return;
+    }
+    const disableNotifications = fcmToken.length == 0;
+
+    const client = {
+      uuid: whoClientId,
+      token: fcmToken,
+      disableNotifcations: disableNotifications,
+      platform: platform,
+      isoCountryCode: isoCountryCode,
+      subscribedTopics: [], // TODO: fill in.
+    } as Client;
+
+    // This update will trigger the `clientSettingsUpdated` method below,
+    // which will actually register (or deregister) the client for notifications.
+    db.collection("Clients").doc(whoClientId).set(client);
+
+    response.status(200).send({});
+  });
+
+// Method that runs when client settings have been updated, and will make those
+// updated settings take effect.
+export const clientSettingsUpdated = functions
+  .region("europe-west6")
```

Comment:
Extract as common variable? Can this be configured at a single place in the terraform code, so that it's simple to change the location of the entire system? Some countries may prefer deploying it within their continent for example.

See single location for the Terraform code:
https://github.com/WorldHealthOrganization/app/blob/master/server/terraform/modules/myhealth/variables.tf#L13

Reply:
We can configure it in a single place in the Firebase code, but it's tricky to share that with Terraform as it currently is. If we care a lot about keeping it in Terraform only, we may find it easier to drop the Firebase tooling for Cloud Functions and use "vanilla" Cloud Functions controlled by a Terraform spec. A bit more work and code, but it would centralize that config value.